src/tangents.jl

```diff
@@ -436,16 +441,75 @@ end
     tangent_field_zeros_exprs = ntuple(fieldcount(P)) do n
         if tangent_field_type(P, n) <: PossiblyUninitTangent
             V = PossiblyUninitTangent{tangent_type(fieldtype(P, n))}
-            return :(isdefined(x, $n) ? $V(zero_tangent(getfield(x, $n))) : $V())
+            return :(isdefined(x, $n) ? $V(zero_tangent_internal(getfield(x, $n))) : $V())
         else
-            return :(zero_tangent(getfield(x, $n)))
+            return :(zero_tangent_internal(getfield(x, $n)))
         end
     end
     backing_data_expr = Expr(:call, :tuple, tangent_field_zeros_exprs...)
     backing_expr = :($(backing_type(P))($backing_data_expr))
     return :($(tangent_type(P))($backing_expr))
 end
 
+@inline zero_tangent_internal(x::Union{Int8,Int16,Int32,Int64,Int128,IEEEFloat}, ::IdDict) = zero_tangent_internal(x)
+@inline function zero_tangent_internal(x::SimpleVector, stackdict::IdDict)
+    return map!(n -> zero_tangent_internal(x[n], stackdict), Vector{Any}(undef, length(x)), eachindex(x))
+end
+@inline function zero_tangent_internal(x::Array{P, N}, stackdict::IdDict) where {P, N}
+    if haskey(stackdict, x)
+        return stackdict[x]::tangent_type(typeof(x))
+    end
+    if isbitstype(P)
+        zt = _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), Array{tangent_type(P), N}(undef, size(x)...), x)
+        stackdict[x] = zt
+        return zt
+    else
+        zt = Array{tangent_type(P), N}(undef, size(x)...)
+        stackdict[x] = zt
+        return _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), zt, x)::Array{tangent_type(P), N}
+    end
```

Comment:
Another thing: do we actually need these two separate paths here? If I'm not mistaken, the second case should cover everything that we need, and be just as performant. I might be wrong though...